### ISSUE TITLE
Missing URL constructor definition

### ISSUE THREAD

**Issue by zuzusik** [Microsoft/TypeScript#2583]
Currently:
```typescript
interface URL {
    revokeObjectURL(url: string): void;
    createObjectURL(object: any, options?: ObjectURLOptions): string;
}
declare var URL: URL;
```
Should be: 
```typescript
interface URLConstructor {
    hash: string;
    search: string;
    pathname: string;
    port: string;
    hostname: string;
    host: string;
    password: string;
    username: string;
    protocol: string;
    origin: string;
    href: string;
}
interface URL {
    revokeObjectURL(url: string): void;
    createObjectURL(object: any, options?: ObjectURLOptions): string;
    new(url: string, base?: string): URLConstructor
}
declare var URL: URL;
```
URL spec is here: https://url.spec.whatwg.org/#api

**Comment by kennysong** [comment]
Where is the URL interface currently defined? I can't find it.

**Comment by mhegazy** [comment]
it is defined in lib.d.ts.

**Comment by kennysong** [comment]
It doesn't seem to be in [lib.d.ts](https://github.com/Microsoft/TypeScript/blob/master/bin/lib.d.ts).

**Comment by danquirk** [comment]
GitHub isn't showing the whole file. It cuts off around 11k lines but URL is defined on line 14455

**Comment by mhegazy** [comment]
look at the raw view: https://raw.githubusercontent.com/Microsoft/TypeScript/master/bin/lib.d.ts

**Comment by myitcv** [comment]
@mhegazy - is this simply a case of modifying the `lib.d.ts` type definition? In which case, are you accepting PRs?

**Comment by myitcv** [comment]
Actually @zhengbli I see your [recent comment](https://github.com/Microsoft/TypeScript/issues/3753#issuecomment-119321813)

Just out of interest, in general how do you determine support within browsers/Javascript VMs? Is there a definitive reference? Your comment in this instance seems to suggest the check was 'by hand'.

**Comment by zuzusik** [comment]
Here is the living standard for it https://url.spec.whatwg.org/#constructors

**Comment by zhengbli** [comment]
@myitcv the `lib.d.ts` is generated from a spec file produced by Edge, and we make manual changes if something is not quite right with the spec or it is Microsoft specific. We are trying to keep the changes as compatible as possible. There is not a "definitive reference" per se, it is all open to discussion.

**Comment by myitcv** [comment]
@zhengbli - thanks, however I was asking about something slightly different. Specifically this comment:

> after trying out on various browsers, we found that the constructor of interface `URL` is not supported in both IE and Spartan

Reading this it sounds like the tests were carried out by hand. I was wondering if in fact there is a definitive reference you use for what functions/etc are supported in each browser?

The [MDN pages](https://developer.mozilla.org/en/docs/Web/API/URL) have a browser compatibility section, but I don't know how up-to-date/accurate these are.

Incidentally, what is "Edge"?

**Comment by zhengbli** [comment]
Oh, for constructors I just open the javascript console in the browser, and run something like `new URL()` and see what happens. If it gives me an error message, so apparently it is not supported.

Edge is the new Microsoft browser coming with windows 10.

**Comment by zuzusik** [comment]
URL constructor is a living standard already. We use it in our code together with polyfill for IE. Is this a reason for not adding it to `lib.d.ts` that it is not yet supported in IE?

**Comment by weswigham** [comment]
Specifically, [this spec](https://url.spec.whatwg.org/). (And its public domain [polyfill](https://github.com/Polymer/URL).) It's an exposure of the internals used to drive the browser's File/Blob URL APIs. Chrome and Firefox implement the living standard pretty well, and it turns out to be really useful to have a URL parser when working on the web. Much neater than using an `img` or `a` tag, anyway.

**Comment by myitcv** [comment]
@weswigham agreed.

**Comment by zhengbli** [comment]
@weswigham @zuzusik using polyfill is a fair point. And as adding a constructor and instance methods is not a breaking change, it might be fine to do that. I'll send a PR to add them. FYI I'm working on open source the script to generate lib.d.ts soon, so later changes could be accepted on that repo.

**Comment by mhegazy** [comment]
Now that the script is to generate the library is open source, PRs are welcomed. Here is some documentation for adding new types to lib.d.ts: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes

**Comment by RReverser** [comment]
Btw, Edge now support URL constructor - does it mean `lib.d.ts` can be just regenerated from Edge spec as stated above?

**Comment by mhegazy** [comment]
We refresh the definition file from that of IE/Edge every few month. we have not done this in a while. @zhengbli has something in the pipeline, not sure if it includes this change though.

**Comment by RReverser** [comment]
Hmm... wouldn't it be possible to somehow automate this? For example, auto-generate `lib.d.ts` from WebIDLs?

**Comment by mhegazy** [comment]
This is what is happening in https://github.com/Microsoft/TSJS-lib-generator. PRs are always welcomed to make the generation more efficient/standard-compliment.

**Comment by RReverser** [comment]
Oh cool, thanks for the link to the repo! But as for "we refresh the definition every few months" - is there any chance to automate this part (as pre-release script for TypeScript?) so that `lib.d.ts` would be always up-to-date?

**Comment by mhegazy** [comment]
These files come from IE/Edge. so we do it when they have a new version available; so we have little control over that. @zhengbli can talk more about the specifics of the next update.. i think he already picked up a new version.

**Comment by zhengbli** [comment]
We are taking another updates after 1.8, as it is very likely that the changes will be big and break many places. Automating the process is definitely possible, however as the spec files are generated from internal builds of Edge, the APIs may not be available yet in the released version. Therefore it makes more sense to review the changes manually now and reduce the interval of updating the spec files in the future.

**Comment by RReverser** [comment]
> however as the spec files are generated from internal builds of Edge, the APIs may not be available yet in the released version

I don't really see how is that a problem - just because public release of Edge doesn't support some new spec APIs natively, that doesn't mean people don't use those APIs (through polyfills, in other browsers, etc.) and doesn't mean they don't want those definitions as soon as they're available... Or am I missing something?

**Comment by mhegazy** [comment]
@RReverser, I am not sure i understand the problem or the proposal here. 

We have lib.d.ts that is generated from the IE/Edge IDL files. @zhengbli updates they, usually quertelly/semi anally. 
These definition files may not contain all APIs, or include addtional non-standard ones, or ones that are not standard-complient, for that, we have three files: https://github.com/Microsoft/TSJS-lib-generator/blob/master/inputfiles/overridingTypes.json, https://github.com/Microsoft/TSJS-lib-generator/blob/master/inputfiles/addedTypes.json and https://github.com/Microsoft/TSJS-lib-generator/blob/master/inputfiles/removedTypes.json that we use to make the output correct.

The PR to add URL constructor for instance is using these fiels, see: https://github.com/Microsoft/TSJS-lib-generator/pull/44.

so am i missing something?

**Comment by RReverser** [comment]
> or include addtional non-standard ones, or ones that are not standard-complient

Gotcha - this was the part I wasn't aware about and thus asked 

> Or am I missing something?

(I thought it's just that internal builds contain APIs that are not released yet, but are fully spec-compliant and still could be used for TS)

**Comment by NekR** [comment]
I am on TS `1.8.9` and this is almost April now, haven't you generated new version yet?

**Comment by mhegazy** [comment]
@NekR, PRs are welcomed. see more information in https://github.com/Microsoft/TypeScript/issues/2583#issuecomment-171478734

**Comment by zhengbli** [comment]
As an update, this issue should be fixed by https://github.com/Microsoft/TypeScript/pull/7619
We recently took a new drop of the XML spec which has the correct definitions for URL.

**Comment by NekR** [comment]
Great, thanks for the update.

**Comment by pritambaral** [comment]
@zhengbli When may we see the fix? The latest release [1.8.10 doesn't have it.](https://github.com/Microsoft/TypeScript/blob/v1.8.10/lib/lib.dom.d.ts#L10969)

**Comment by zhengbli** [comment]
It is already in master: https://github.com/Microsoft/TypeScript/blob/35b8b42b554d2b98230320b727b452d0abe9aaae/src/lib/dom.generated.d.ts#L11740

it should be in next release.
